aptos-move/move-examples/fungible_asset/coin_wrapper/sources/coin_wrapper.move

```diff
@@ -0,0 +1,220 @@
+        let burn_ref = &fungible_asset_data<CoinType>().burn_ref;
+        fungible_asset::burn(burn_ref, fa);
+        let wrapper_signer = &account::create_signer_with_capability(&wrapper_account().signer_cap);
+        coin::withdraw(wrapper_signer, amount)
+    }
+
+    /// Create the fungible asset wrapper for the given CoinType if it doesn't exist yet.
+    public(friend) fun create_fungible_asset<CoinType>(): Object<Metadata> acquires WrapperAccount {
+        let coin_type = type_info::type_name<CoinType>();
+        let wrapper_account = mut_wrapper_account();
+        let coin_to_fungible_asset = &mut wrapper_account.coin_to_fungible_asset;
+        let wrapper_signer = &account::create_signer_with_capability(&wrapper_account.signer_cap);
+        if (!smart_table::contains(coin_to_fungible_asset, coin_type)) {
+            let metadata_constructor_ref = &object::create_named_object(wrapper_signer, *string::bytes(&coin_type));
+            primary_fungible_store::create_primary_store_enabled_fungible_asset(
+                metadata_constructor_ref,
+                option::none(),
+                coin_type,
+                coin::symbol<CoinType>(),
+                coin::decimals<CoinType>(),
+                string::utf8(b""),
+                string::utf8(b""),
+            );
+
+            let mint_ref = fungible_asset::generate_mint_ref(metadata_constructor_ref);
+            let burn_ref = fungible_asset::generate_burn_ref(metadata_constructor_ref);
+            let metadata = object::object_from_constructor_ref<Metadata>(metadata_constructor_ref);
+            smart_table::add(coin_to_fungible_asset, coin_type, FungibleAssetData {
+                metadata,
+                mint_ref,
+                burn_ref,
+            });
+            smart_table::add(&mut wrapper_account.fungible_asset_to_coin, metadata, coin_type);
+        };
+        smart_table::borrow(coin_to_fungible_asset, coin_type).metadata
+    }
+
+    inline fun fungible_asset_data<CoinType>(): &FungibleAssetData acquires WrapperAccount {
+        let coin_type = type_info::type_name<CoinType>();
+        smart_table::borrow(&wrapper_account().coin_to_fungible_asset, coin_type)
+    }
+
+    inline fun wrapper_account(): &WrapperAccount acquires WrapperAccount {
+        borrow_global<WrapperAccount>(wrapper_address())
+    }
+
+    inline fun mut_wrapper_account(): &mut WrapperAccount acquires WrapperAccount {
+        borrow_global_mut<WrapperAccount>(wrapper_address())
+    }
+
+    #[test_only]
+    use aptos_framework::coin::FakeMoney;
+    #[test_only]
+    use std::signer;
+
+    #[test_only]
+    public fun initialize_for_test(deployer: &signer) {
+        init_module(deployer);
+    }
+
+    #[test(user = @0x1, deployer = @0xcafe)]
+    fun e2e_test(user: &signer, deployer: &signer) acquires WrapperAccount {
+        account::create_account_for_test(signer::address_of(user));
+        account::create_account_for_test(signer::address_of(deployer));
```

Comment:
nit: can simplify into

```
#[test]
fun e2e_test() acquires WrapperAccount {
  let user = account::create_account_for_test(@0x1);
  let deployer = account::create_account_for_test(@0xcafe);
  ...
}
```

Reply:
I like having the accounts declared at the top of the test. Probably just individual preferences :)

Reply:
the more normalized, the better.